packages/plugins/plugin-assistant/src/components/QueryEditor/query.test.ts

```diff
@@ -0,0 +1,254 @@
+// TODO(burdon): Ref/Relation traversal.
+// TODO(burdon): Factor out.
+
+describe('query', () => {
+  it('should parse a simple query', ({ expect }) => {
+    const queryParser = parser.configure({ strict: true });
+
+    const tests = [
+      {
+        query: 'type:dxos.org/type/Contact',
+        parts: [
+          'Query',
+          // type:dxos.org/type/Contact
+          'Filter',
+          'TypeFilter',
+          'TypeKeyword',
+          ':',
+          'Identifier',
+        ],
+      },
+      {
+        query: '{ name: "DXOS" }',
+        parts: [
+          'Query',
+          // { name: "DXOS" }
+          'Filter',
+          'ObjectLiteral',
+          '{',
+          'ObjectProperty',
+          'PropertyKey',
+          'Identifier',
+          ':',
+          'Value',
+          'String',
+          '}',
+        ],
+      },
+      {
+        query: '{ value: 100 }',
+        parts: [
+          'Query',
+          // { value: 100 }
+          'Filter',
+          'ObjectLiteral',
+          '{',
+          'ObjectProperty',
+          'PropertyKey',
+          'Identifier',
+          ':',
+          'Value',
+          'Number',
+          '}',
+        ],
+      },
+      {
+        query: '{ value: true }',
+        parts: [
+          'Query',
+          // { value: true }
+          'Filter',
+          'ObjectLiteral',
+          '{',
+          'ObjectProperty',
+          'PropertyKey',
+          'Identifier',
+          ':',
+          'Value',
+          'Boolean',
+          '}',
+        ],
+      },
+      {
+        query: 'type:dxos.org/type/Contact OR type:dxos.org/type/Organization',
+        parts: [
+          'Query',
+          // type:dxos.org/type/Contact
+          'Filter',
+          'TypeFilter',
+          'TypeKeyword',
+          ':',
+          'Identifier',
+          // OR
+          'Or',
+          // type:dxos.org/type/Organization
+          'Filter',
+          'TypeFilter',
+          'TypeKeyword',
+          ':',
+          'Identifier',
+        ],
+      },
+      {
+        query: '(type:dxos.org/type/Contact OR type:dxos.org/type/Organization) AND { name: "DXOS" }',
+        parts: [
+          'Query',
+          '(',
+          // type:dxos.org/type/Contact
+          'Filter',
+          'TypeFilter',
+          'TypeKeyword',
+          ':',
+          'Identifier',
+          // OR
+          'Or',
+          // type:dxos.org/type/Organization
+          'Filter',
+          'TypeFilter',
+          'TypeKeyword',
+          ':',
+          'Identifier',
+          ')',
+          // AND
+          'And',
+          // { name: "DXOS" }
+          'Filter',
+          'ObjectLiteral',
+          '{',
+          'ObjectProperty',
+          'PropertyKey',
+          'Identifier',
+          ':',
+          'Value',
+          'String',
+          '}',
+        ],
+      },
+      {
+        query: '(type:dxos.org/echo/Contact) => (type:dxos.org/echo/Organization)',
```

Comment:
not sure how I should interpret this, which relation does this traverse? what about traversing reference fields?

Reply:
i haven't figured out a representation for refs or constrained relations yet